yarn-project/simulator/src/public/public_tx_simulator.ts

```diff
@@ -0,0 +1,263 @@
+import { type DebugLogger, createDebugLogger } from '@aztec/foundation/log';
+import { Timer } from '@aztec/foundation/timer';
+
+import { EnqueuedCallSimulator } from './enqueued_call_simulator.js';
+import { type PublicExecutor } from './executor.js';
+import { type WorldStateDB } from './public_db_sources.js';
+import { type PublicKernelCircuitSimulator } from './public_kernel_circuit_simulator.js';
+import { PublicKernelTailSimulator } from './public_kernel_tail_simulator.js';
+import { PublicTxContext } from './public_tx_context.js';
+import { generateAvmCircuitPublicInputs, runMergeKernelCircuit } from './utils.js';
+
+export type ProcessedPhase = {
+  phase: TxExecutionPhase;
+  durationMs: number;
+  returnValues: NestedProcessReturnValues[];
+  reverted: boolean;
+  revertReason?: SimulationError;
+};
+
+export type PublicTxResult = {
+  avmProvingRequest: AvmProvingRequest;
+  /** Gas used during the execution of this tx */
+  gasUsed: GasUsed;
+  revertCode: RevertCode;
+  /** Revert reason, if any */
+  revertReason?: SimulationError;
+  processedPhases: ProcessedPhase[];
+};
+
+export class PublicTxSimulator {
+  private log: DebugLogger;
+
+  constructor(
+    private db: MerkleTreeReadOperations,
+    private publicKernelSimulator: PublicKernelCircuitSimulator,
+    private globalVariables: GlobalVariables,
+    private worldStateDB: WorldStateDB,
+    private enqueuedCallSimulator: EnqueuedCallSimulator,
+    private publicKernelTailSimulator: PublicKernelTailSimulator,
+  ) {
+    this.log = createDebugLogger(`aztec:sequencer`);
+  }
+
+  static create(
+    db: MerkleTreeReadOperations,
+    publicExecutor: PublicExecutor,
+    publicKernelSimulator: PublicKernelCircuitSimulator,
+    globalVariables: GlobalVariables,
+    historicalHeader: Header,
+    worldStateDB: WorldStateDB,
+    realAvmProvingRequests: boolean = true,
+  ) {
+    const enqueuedCallSimulator = new EnqueuedCallSimulator(
+      db,
+      worldStateDB,
+      publicExecutor,
+      globalVariables,
+      historicalHeader,
+      realAvmProvingRequests,
+    );
+
+    const publicKernelTailSimulator = PublicKernelTailSimulator.create(db, publicKernelSimulator);
+
+    return new PublicTxSimulator(
+      db,
+      publicKernelSimulator,
+      globalVariables,
+      worldStateDB,
+      enqueuedCallSimulator,
+      publicKernelTailSimulator,
+    );
+  }
+
+  async process(tx: Tx): Promise<PublicTxResult> {
+    this.log.verbose(`Processing tx ${tx.getTxHash()}`);
+
+    const context = await PublicTxContext.create(this.db, this.worldStateDB, tx, this.globalVariables);
+
+    const setupResult = await this.processSetupPhase(context);
+    const appLogicResult = await this.processAppLogicPhase(context);
+    const teardownResult = await this.processTeardownPhase(context);
+
+    const processedPhases = [setupResult, appLogicResult, teardownResult].filter(
+      result => result !== undefined,
```

Comment:
Could you comment (in the code) why a state could be undefined and what that means?